internal/db/update/update.go

```diff
@@ -21,76 +18,85 @@ CREATE TABLE schemas (
 );
 `
 
-// Template for schema files (can't use backticks since we need to use backticks
-// inside the template itself).
-const dotGoTemplate = "package %s\n\n" +
-	"// DO NOT EDIT BY HAND\n" +
-	"//\n" +
-	"// This code was generated by the schema.DotGo function. If you need to\n" +
-	"// modify the database schema, please add a new schema update to update.go\n" +
-	"// and the run 'make update-schema'.\n" +
-	"const freshSchema = `\n" +
-	"%s`\n"
-
 type SchemaUpdateManager struct {
-	updates map[int]schema.Update
+	updates map[updateType][]schema.Update
 }
 
 func NewSchema() *SchemaUpdateManager {
 	return &SchemaUpdateManager{
-		updates: map[int]schema.Update{
-			1: updateFromV0,
+		updates: map[updateType][]schema.Update{
+			updateInternal: {
+				updateFromV0,
+				updateFromV1,
+			},
 		},
 	}
 }
 
 func (m *SchemaUpdateManager) Schema() *SchemaUpdate {
-	schema := NewFromMap(m.updates)
+	schema := &SchemaUpdate{updates: m.updates}
 	schema.Fresh("")
 	return schema
 }
 
-func (m *SchemaUpdateManager) AppendSchema(extensions map[int]schema.Update) {
-	currentVersion := len(m.updates)
-	schema := NewFromMap(extensions)
-	for _, extension := range schema.updates {
-		m.updates[currentVersion+1] = extension
-		currentVersion = len(m.updates)
-	}
+func (m *SchemaUpdateManager) AppendSchema(extensions []schema.Update) {
+	m.updates[updateExternal] = extensions
 }
 
-func (m *SchemaUpdateManager) SchemaDotGo() error {
-	// Apply all the updates that we have on a pristine database and dump
-	// the resulting schema.
-	db, err := sql.Open("sqlite3", ":memory:")
-	if err != nil {
-		return fmt.Errorf("failed to open schema.go for writing: %w", err)
-	}
-
-	schema := NewFromMap(m.updates)
+// updateFromV1 fixes a bug in the schemas table. Previously there was no way to tell when an update was internal, so the last external update would be re-run instead.
+// To fix this, we introduce a `type` column to the schemas table. The first schema update will be considered internal, as that is what microcluster initially shipped with.
+// All other schema versions will be considered external.
+//
+// Because this update affects the schema update mechanism, it must necessarily be run manually before the regular schema updates.
+// So first it checks if the column exists already, and in such a case does nothing.
+func updateFromV1(ctx context.Context, tx *sql.Tx) error {
```

Comment:
I think as this schema update is quite critical it would be worth having a unit test to verify that it does exactly what is expected.